File: arrays/jump_game_ii_bfs.py
from collections import deque

def jump(nums)
    if len(nums) <= 1:
        return 0

    queue = deque([(0, 0)])
    visited = set([0])

    while queue:
        index, jumps = queue.popleft()
        max_jump = min(index + nums[index], len(nums) - 1)
        for next_index in range(index + 1, max_jump + 1):
            if next_index == len(nums) - 1:
                return jumps + 1
            if next_index not in visited:
                queue.append((next_index, jumps + 1))
                visited.add(next_index)

File: arrays/jump_game_ii_bfs.rb
require 'set'

def jump(nums)
    return 0 if nums.size <= 1

    queue = [[0,0]]
    visited = Set.new([0])

    while !queue.empty?
        index, jumps = queue.shift

        max_jump = [index + nums[index], nums.size - 1].min
        ((index + 1)..max_jump).each do |next_index|
            queue << [next_index, jumps + 1]
            visited << next_index
        end
    end
end